auth/provider.go

```diff
@@ -56,26 +61,280 @@ func (r *NoAuthScope) Allow(_ string, _ string) (ok bool) {
 	client := gocloak.NewClient(host)
 	k = Keycloak{
-		host:   host,
-		realm:  realm,
-		id:     id,
-		secret: secret,
-		client: client,
+		host:       host,
+		realm:      realm,
+		id:         id,
+		secret:     secret,
+		client:     client,
+		admin:      admin,
+		pass:       pass,
+		adminRealm: adminRealm,
 	}
 	return
 }
 
 //
 // Keycloak auth provider
 type Keycloak struct {
-	client gocloak.GoCloak
-	host   string
-	realm  string
-	id     string
-	secret string
+	client     gocloak.GoCloak
+	host       string
+	realm      string
+	id         string
+	secret     string
+	admin      string
+	pass       string
+	adminRealm string
+	token      *gocloak.JWT
+}
+
+//
+// Realm is a container for the users,
+// scopes, and roles that exist in the
+// hub's keycloak realm.
+type Realm struct {
+	Users  map[string]gocloak.User
+	Scopes map[string]gocloak.ClientScope
+	Roles  map[string]gocloak.Role
+}
+
+//
+// Reconcile ensures that Keycloak is configured
+// with the scopes, roles, and users required by
+// the hub.
+func (r *Keycloak) Reconcile() (err error) {
+	err = r.login()
+	if err != nil {
+		return
+	}
+
+	realm, err := r.loadRealm()
+	if err != nil {
+		return
+	}
+
+	err = r.ensureRoles(realm)
+	if err != nil {
+		return
+	}
+
+	err = r.ensureUsers(realm)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+//
+// loadRealm loads all the scopes, roles, and users from
+// the hub keycloak realm and populates a Realm struct.
+func (r *Keycloak) loadRealm() (realm *Realm, err error) {
+	realm = &Realm{}
+	realm.Scopes, err = r.scopeMap()
+	if err != nil {
+		return
+	}
+	realm.Roles, err = r.realmRoleMap()
+	if err != nil {
+		return
+	}
+	realm.Users, err = r.userMap()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+//
+// ensureUsers ensures that the hub users exist and have the necessary roles.
+func (r *Keycloak) ensureUsers(realm *Realm) (err error) {
+	users, err := LoadUsers(Settings.Auth.UserPath)
+	if err != nil {
+		return
+	}
+
+	for _, user := range users {
+		if _, ok := realm.Users[user.Name]; !ok {
+			_true := true
+			u := gocloak.User{
+				Username:        &user.Name,
+				Enabled:         &_true,
+				RequiredActions: &[]string{"UPDATE_PASSWORD"},
+			}
+			userid, kErr := r.client.CreateUser(context.Background(), r.token.AccessToken, r.realm, u)
+			if kErr != nil {
+				err = liberr.Wrap(kErr)
+				return
+			}
+			u.ID = &userid
+			err = r.client.SetPassword(context.Background(), r.token.AccessToken, userid, r.realm, user.Password, true)
+			if err != nil {
+				err = liberr.Wrap(err)
+				return
+			}
+			realm.Users[user.Name] = u
+		}
+		var realmRoles []gocloak.Role
+		for _, role := range user.Roles {
+			realmRole, ok := realm.Roles[role]
+			if !ok {
+				continue
+			}
+			realmRoles = append(realmRoles, realmRole)
+		}
+		err = r.client.AddRealmRoleToUser(context.Background(), r.token.AccessToken, r.realm, *realm.Users[user.Name].ID, realmRoles)
```

Comment:
Will this fail if the user already has one or more of the roles?

Reply:
No, this does not appear to fail.